### ISSUE TITLE
Allow creating complex key condition expressions

### ISSUE THREAD

**Issue by louislatreille** [sensedeep/dynamodb-onetable#431]
Currently, using find, I can only use one operation against the sort key. It would be awesome to be able to build more complex expressions.

Example code:
```
await mainTable.getModel("workoutSessionEvent").find(
  {
    workoutSessionId,
    SK: { ">": `WE#${eventId}`, begins_with: "WE#" },
  },
  {
    postFormat: (model, args) => {
      console.log(model);
      console.log(args);

      return args;
    },
  },
)
```

Logs:
```
{
  ExpressionAttributeNames: { '#_0': 'PK', '#_1': 'SK', '#_2': 'workoutSessionId' },
  ExpressionAttributeValues: {
    ':_0': { S: 'WS#2JKk3PbehdqUF9IVUniYxNiSZtr' },
    ':_1': { S: 'WE#16712095501963541840006' },
    ':_2': { S: '2JKk3PbehdqUF9IVUniYxNiSZtr' }
  },
  FilterExpression: '#_2 = :_2',
  KeyConditionExpression: '#_0 = :_0 and #_1 > :_1',
  TableName: '######',
  ConsistentRead: false,
  ScanIndexForward: true
}
```

I would have expected another `and` in the key condition expression above, something like:
```
{
  ExpressionAttributeNames: { '#_0': 'PK', '#_1': 'SK', '#_2': 'workoutSessionId' },
  ExpressionAttributeValues: {
    ':_0': { S: 'WS#2JKk3PbehdqUF9IVUniYxNiSZtr' },
    ':_1': { S: 'WE#16712095501963541840006' },
    ':_2': { S: '2JKk3PbehdqUF9IVUniYxNiSZtr' },
    ':_3': { S: 'WE#' },
  },
  FilterExpression: '#_2 = :_2',
  KeyConditionExpression: '#_0 = :_0 and #_1 > :_1 and begins_with(#_1, :_3)',
  TableName: '######',
  ConsistentRead: false,
  ScanIndexForward: true
}
```

**Comment by mobsense** [comment]
You can with where clauses!

https://doc.onetable.io/api/model/where/

```
let adminUsers = await User.find({}, {
    where: '(${role} = {admin}) and (${status} = @{status})',
    substitutions: {
        status: 'current'
    }
})

```

**Comment by louislatreille** [comment]
Thanks for your response @mobsense !

I tried with a where clause today, and it doesn't see to work as expected. I'm even getting an error with it.

Here's my updated code:
```
await mainTable.getModel("workoutSessionEvent").find(
  {
    workoutSessionId,
  },
  {
    where: "(${SK} > @{skGreaterThan}) and (begins_with(${SK}, {WE#}))",
    substitutions: {
      skGreaterThan: `WE#${eventId}`,
    },
    postFormat: (_, args) => {
      console.log(args);
      return args;
    },
  },
})
```

This gives the following DynamoDB query:
```
{
  ExpressionAttributeNames: { '#_0': 'SK', '#_1': 'PK', '#_2': 'workoutSessionId' },
  ExpressionAttributeValues: {
    ':_0': { S: 'WE#16731960678297225260000' },
    ':_1': { S: 'WE#' },
    ':_2': { S: 'WS#2K3GctDzM0H5F5tjPm926BcQ5qG' },
    ':_3': { S: '2K3GctDzM0H5F5tjPm926BcQ5qG' }
  },
  FilterExpression: '(#_0 > :_0 and begins_with(#_0, :_1)) and (#_2 = :_3)',
  KeyConditionExpression: '#_1 = :_2 and begins_with(#_0, :_1)',
  TableName: 'backend-alpha-ca-central-1-persistence-table',
  ConsistentRead: false,
  ScanIndexForward: true
}
```

The `KeyConditionExpression` is missing the greater than, and the `FilterExpression` contains all the SK conditions but it shouldn't contain any. The latter is actually causing an error with DynamoDB:

```
ERROR	fe: OneTable execute failed "find" for "workoutSessionEvent. Filter Expression can only contain non-primary key attributes: Primary key attribute: SK
    at yl.execute (/var/task/index.js:65:57768)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Wn.run (/var/task/index.js:65:14354)
    at async Wn.queryItems (/var/task/index.js:65:24222)
    at async Wn.find (/var/task/index.js:65:18202)
    at async Object.getWorkoutEventsNewerThan (/var/task/index.js:68:87792)
    at async /var/task/index.js:68:98939
    at async ED (/var/task/index.js:42:8100) {
  context: {
    err: ValidationException: Filter Expression can only contain non-primary key attributes: Primary key attribute: SK
        at bt (/var/task/index.js:68:1525)
        at C8 (/var/task/index.js:68:28592)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async /var/task/index.js:66:2310
        at async /var/task/index.js:61:2577
        at async Gs.retry (/var/task/index.js:50:16934)
        at async /var/task/index.js:50:11344
        at async bm.send (/var/task/index.js:68:77121)
        at async bm.find (/var/task/index.js:68:76600)
        at async yl.execute (/var/task/index.js:65:57165) {
      '$fault': 'client',
      '$metadata': [Object],
      __type: 'com.amazon.coral.validate#ValidationException'
    },
    trace: {
      model: 'workoutSessionEvent',
      cmd: [Object],
      op: 'find',
      properties: [Object]
    }
  },
```

**Comment by dev-embedthis** [comment]
Can you post a reproducible test case that we can use to dig deeper.

We need your model and code as per the issue template so we can diagnose.

OneTable should separate those into keys and filter expressions and we use that pattern ourselves in many cases.

**Comment by dmeehan1968** [comment]
I have been experiencing a similar problem, in that a `FilterExpression` is generated for the fields specified as part of properties, even though they have been used in a value expression.  

For me, I thought it was because I'm using a `value: true` in the schema, and providing a value function via the Table params (in which case, I suspected OneTable wouldn't be able to determine that the properties was used in the value expression).  But I tried simplifying the model and value expression to a string, and that still generated a filter expression.

All of the docs examples/samples seem to use `pk` and `sk` directly, do a 'find all', or on an attribute that won't result in a match due to partitioning (some finds aren't validated), and/or there are no value templates specified in the models,and certainly no value functions, so I don't think the docs adequately show how this should be done/if it works in practice, beyond very simple illustrations.  

Looking at the tests for `find` these seem to be overly naive and wouldn't pick up whether the generated query contains an unexpected filter expression that happened to also match the index definition, leading to an expected match in item count.  So there should definitely be tests that check for the possibility (maybe there are tests elsewhere that validate the command input, I didn't look).  There should also be some tests when value functions are used (its not clear how OneTable determines whether the supplied property should be used as a filter if its also used in a value template).

Model: 

```ts
Debug: {
    PK: { type: String, value: true },
    SK: { type: String, value: 'DEBUG#${deviceTime}' },
}
```

Table/Value function:

NB: The deviceTime is truncated to create hourly shards in my case, and is the reason for the value function as a string can only substitute the full value.

```ts
new Table({ 
  client, 
  name, 
  schema, 
  value: (props) => `DEV#${properties.devId}#STATE#${properties.deviceTime?.toISOString().slice(0, 7)}`,
})
```

Find:

```ts
await Debug.find({
  devId: "01GQ38T0HDJKZ2AW4M0D78Z7V8",
  deviceTime: new Date("2023-01-25T12:00:00.000Z"),
})
```

As best as I can figure out from the docs, this should create a PK value based on the provided properties (so should call the value function to derive PK), and the SK should be a partial match (`DEBUG#`).  Example strings, (not the actual expressions sent to DynamoDB):

```
PK = DEV#01GQ38T0HDJKZ2AW4M0D78Z7V8#DEBUG#2023-01-25T12
SK begins DEBUG#
```

What this generates is:

```
Error: Unsupported type passed: Wed Jan 25 2023 12:00:00 GMT+0000 (Coordinated Universal Time). Pass options.convertClassInstanceToMap=true to marshall typeof object as map attribute.
   at convertToAttr (/[redacted]/amplify/backend/function/loriotIntegration/src/node_modules/@aws-sdk/util-dynamodb/dist-cjs/convertToAttr.js:49:11)

```

This appears to be passing the deviceTime property to the DynamoDB client instance (which seems to be an error in handling the property, because it should probably have been converted to an isoDate string. I can mute it with:

```ts
new Dynamo({ client, marshall: {convertClassInstanceToMap: true })
```

The result is then:

```
info OneTable result for "find" "Debug" {
    "cmd": {
        "ExpressionAttributeNames": {
            "#_0": "PK",
            "#_1": "SK",
            "#_2": "devId",
            "#_3": "deviceTime"
        },
        "ExpressionAttributeValues": {
            ":_0": {
                "S": "DEV#01GQ38T0HDJKZ2AW4M0D78Z7V8#DEBUG#2023-01-25T12"
            },
            ":_1": {
                "S": "DEBUG#2023-01-25T12:00:00.000Z"
            },
            ":_2": {
                "S": "01GQ38T0HDJKZ2AW4M0D78Z7V8"
            },
            ":_3": {
                "M": {}
            }
        },
        "FilterExpression": "(#_2 = :_2) and (#_3 = :_3)",
        "KeyConditionExpression": "#_0 = :_0 and #_1 = :_1",
        "TableName": "redactedTable",
        "ReturnConsumedCapacity": "INDEXES",
        "ReturnItemCollectionMetrics": "SIZE",
        "ConsistentRead": false,
        "ScanIndexForward": true
    },
    "items": [],
    "op": "find",
    "properties": {
        "PK": "DEV#01GQ38T0HDJKZ2AW4M0D78Z7V8#DEBUG#2023-01-25T12",
        "SK": "DEBUG#2023-01-25T12:00:00.000Z",
        "devId": "01GQ38T0HDJKZ2AW4M0D78Z7V8",
        "deviceTime": "2023-01-25T12:00:00.000Z"
    },
    "params": {
        "parse": true,
        "high": true,
        "log": true,
        "checked": true
    }
}
```

The only way I could get it to work was to specify PK and SK properties, and modify the value function to return those values if they are already in the properties:

```ts
new Table({ 
  client, 
  name, 
  schema, 
  value: (props) => properties.PK ? `DEV#${properties.devId}#STATE#${properties.deviceTime?.toISOString().slice(0, 7)}`,
})
```

```ts
await Debug.find({
  PK: "DEV#01GQ38T0HDJKZ2AW4M0D78Z7V8#DEBUG#2023-01-25T12",
  SK: { begins: "DEBUG#" },
})
```

This results in:

```
info OneTable result for "find" "Debug" {
    "cmd": {
        "ExpressionAttributeNames": {
            "#_0": "PK",
            "#_1": "SK"
        },
        "ExpressionAttributeValues": {
            ":_0": {
                "S": "DEV#01GQ38T0HDJKZ2AW4M0D78Z7V8#DEBUG#2023-01-25T12"
            },
            ":_1": {
                "S": "DEBUG#"
            }
        },
        "KeyConditionExpression": "#_0 = :_0 and begins_with(#_1, :_1)",
        "TableName": "wavetrendTable-dev",
        "ReturnConsumedCapacity": "INDEXES",
        "ReturnItemCollectionMetrics": "SIZE",
        "ConsistentRead": false,
        "ScanIndexForward": true
    },
    "items": [ /* redacted */ ],
    "op": "find",
    "properties": {
        "PK": "DEV#01GQ38T0HDJKZ2AW4M0D78Z7V8#DEBUG#2023-01-25T12",
        "SK": {
            "begins": "DEBUG#"
        }
    },
    "params": {
        "parse": true,
        "high": true,
        "log": true,
        "checked": true
    }
}
```

Sorry, I don't really have the time to generate a specific runnable test case for this.

**Comment by dmeehan1968** [comment]
I realise now that I had completely the wrong end of the stick about how `find()` worked, a combination of putting trust in type safety and the use of value functions to construct the primary (and/or) other keys.

Initially I assumed that by suppling the properties that make up the primary keys would be sufficient for OneTable to determine the keys.  Because I was using value functions, I assumed (my bad) that it was relying on the use of known properties in string type value templates to determine which properties were supplied as key components and which were properties intended for filter expressions.

My feeling is that the conflation of the index properties and filter properties makes this hard to determine, and properties argument type doesn't enforce 'at least one' of the index values allows for arbitrary properties to be supplied, but instead seems to make them all optional (via Partial).  `find()` without an index property might well be translated into a `scan` but that seesm dangerous given the potential time/cost of unintentionally invoking a scan.

What I discovered through trial and error was that I needed to supplied the index values, contra to what I'd already been doing with the `create()` calls (where you supply the attributes and not the indexes).  If I wanted a filter expression, that's what the `where` parameter is for (and in my case I didn't want a filter).  I think for me the first error relating to `convertClassInstanceToMap` was due to supplying a `Date` value which wasn't being converted to an ISODate string, again, as is the way with `create()`.

It would have been a lot clearer if the signature for `find()` only took the index values (and ensured no overlap) as the properties argument, and either another argument for properties to be used in filter expressions, or solely support that via `where`.

If the Schema defines the indexes as:

```ts
indexes: {
    primary: { hash: 'PK', sort: 'SK', },
    GS1: { hash: 'GS1PK', sort: 'GS1SK', },
},
```

Then the find properties (for indexes only) should be (via a discriminated union):

```ts
type FindProperties = { PK: string, SK: string, GS1PK: undefined, GS1SK: undefined } | { PK: undefined, SK: undefined, GS1PK: string, GS1SK: string }
```

This would ensure that `find()` was used for Query and not an unintentional scan, and would remove the need to optionally state `params.index` in the case of a non-primary index.  I recognise that part of the difficulty here is that the actual type of the partition/sort key, given that its part of the model and might be part of a single table design.  I think that partition/sort keys can be string or number within DynamoDB (not sure if it allows a combination across items).

Obviously it would be a breaking change to alter the existing treatment of find properties, and probably therefore it is what it is.  Maybe what I'm suggesting here is that the docs/examples are expanded to make it clear the interplay between index values and attribute values and how that translates to key conditions and filter expressions.  

Maybe I just got the mental model wrong having already worked with create(), and the use of value functions already in the design.  I think I also might have hijacked this thread as it appeared related.
